fix: multiple matches in a single line in YAML view filter

shidenkai0 opened the pull request:
# Changes
- Correctly handle multiple matches in a single line when filtering in the YAML view
- Add a unit test for rxFilter

# Demonstration
<img width="2112" alt="Screenshot 2022-11-26 at 19 53 38" src="https://user-images.githubusercontent.com/8296549/204105125-ec74b122-0866-44c0-833c-a1808bf13941.png">
As you can see in the screenshot above (right), the 8 matches of "bar" are correctly detected and can be navigated through, whereas, on the current rev (left), only 2 matches are detected (one per line).

closes #1727

slimus commented on internal/view/live_view.go:
```diff
@@ -91,23 +91,35 @@ func (v *LiveView) ResourceFailed(err error) {
+		loc, line := m.MatchedIndexes, ll[m.Index]
+		offset := offsetForLine[m.Index]
+		loc[0] += offset
+		loc[1] += offset
```
Why for both 0 and 1 locations use the same offset? Can you please add unit tests here?

shidenkai0 replied:
Hi @slimus,
Thanks for the review!
loc[0] and loc[1] represent the start and end of a match, to be quoted into region tags, therefore the same offset applies to both. The goal of this offset is to take into account that when multiple matches occur in the same line, and when inserting region tags around a match, the start and end index of subsequent matches get offset by the length of the region tags.
Regarding the unit tests, I will add some.

slimus replied:
Awesome! Thanks for the explanation and unit tests.

derailed commented on internal/model/yaml_test.go:
```diff
@@ -0,0 +1,67 @@
+				},
+			},
+		},
+		"multiple-matches": {
```
Might be a good idea to add a use cases where the match is contained multiple time in the line and not necessarily space delimited ie "thefoobar fobarbarfofool barfoofo"...

derailed commented on internal/view/live_view.go:
```diff
@@ -98,9 +98,15 @@ func (v *LiveView) ResourceChanged(lines []string, matches fuzzy.Matches) {
 			loc, line := m.MatchedIndexes, ll[m.Index]
-			ll[m.Index] = line[:loc[0]] + `<<<"search_` + strconv.Itoa(i) + `">>>` + line[loc[0]:loc[1]] + `<<<"">>>` + line[loc[1]:]
+			offset := offsetForLine[m.Index]
+			loc[0] += offset
```
I get the use here but better to use parrallel assign ie loc[0], loc[1] = loc[0]+offset,loc[1]+offset. Makes it cleared imho as they are related.

derailed commented on internal/view/live_view.go:
```diff
@@ -98,9 +98,15 @@ func (v *LiveView) ResourceChanged(lines []string, matches fuzzy.Matches) {
+			offset := offsetForLine[m.Index]
+			loc[0] += offset
+			loc[1] += offset
+			regionStr := fmt.Sprintf(`<<<"search_%d">>>%s<<<"">>>`, i, line[loc[0]:loc[1]])
```
Not keen on this in terms of allocs. Prefer the old way (tho not as readable...)

derailed reviewed (changes requested):
@shidenkai0 Thank you for the fix Mohamed! Just a few small picks...

shidenkai0 commented:
> @shidenkai0 Thank you for the fix Mohamed! Just a few small picks...

Thank you for the review @derailed! I pushed a commit including your feedback.

derailed commented on internal/model/yaml_test.go:
```diff
@@ -14,9 +14,8 @@ func TestYAML_rxFilter(t *testing.T) {
-			lines: []string{},
-			e:     fuzzy.Matches{},
+			q: "foo",
+			e: fuzzy.Matches{},
```
Same here this is the zero value for this type.

shidenkai0 replied:
I left this occurrence because the expected slice in this case is an empty slice, while the zero value for this type is a `nil` slice. The alternative would be to not preallocate the slice in the function with `make`, but that would be imho less effective. What do you think?

derailed replied:
Doh! I stand corrected. You are absolutely right Mohamed. Thank you!!

derailed commented on internal/view/live_view_test.go:
```diff
@@ -14,14 +14,11 @@ func TestLiveView_linesWithRegions(t *testing.T) {
-			lines:   []string{},
-			matches: fuzzy.Matches{},
-			e:       []string{},
+			e: []string{},
```
Same here. This is the zero value.

shidenkai0 replied:
Same as above, the zero value here is a `nil` slice.

derailed commented on internal/view/live_view_test.go:
```diff
@@ -39,14 +36,13 @@ func TestLiveView_linesWithRegions(t *testing.T) {
 			e: []string{"foo", `<<<"search_0">>>bar<<<"">>>`, `<<<"search_1">>>baz<<<"">>>`},
```
In the test you could use a formatter helpers ie matchTag(i int, s string) string. Easier to track should syntax changes or for new test cases.

derailed reviewed (changes requested):
@shidenkai0 Thank you for your updates. Looks great!! Just a few more picks. 
Also might be worth squashing the pr review commits as part of the original pr.

shidenkai0 commented:
Thank you for your review @derailed ! I updated the PR and squashed the commits, I left some comments regarding the zero value of the `e` slices in the table test, what do you think?

derailed commented on internal/model/yaml_test.go:
```diff
@@ -0,0 +1,66 @@
+	for k := range uu {
+		u := uu[k]
+		t.Run(k, func(t *testing.T) {
+			y := YAML{}
```
Might want to move this out of the loop.  var y YAML instead

derailed commented on internal/view/live_view_test.go:
```diff
@@ -0,0 +1,64 @@
+			},
+		},
+	}
+	v := LiveView{}
```
Prefer var v LiveView

derailed reviewed (changes requested):
@shidenkai0 Thank you for the updates and corrections!

shidenkai0 commented:
Hi @derailed,

Thanks for the updated review! I implemented your latest feedback and pushed the changes.

derailed reviewed (approved):
@shidenkai0 Awesome! Thank you for this PR and the updates Mohamed!!
